[FLINK-11474][table] Add ReadableCatalog, ReadableWritableCatalog, and other …

xuefuz opened the pull request:
…related interfaces

<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](http://flink.apache.org/contribute-code.html#best-practices).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*Add new catalog APIs pertaining to FLIP-30*


## Brief change log

  - *Add ReadableCatalog and ReadableWritableCatalog interfaces*
  - *Add catalog related interfaces and classes such as CommonTable, CatalogTable, CatalogView, etc.*
  - *Migrated some old exception classes from scala to java.*
  - *Old interfaces or classes remains but are subject to removal once new implementation is in place.*

## Verifying this change

*This change is mainly about interfaces without any test coverage.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
 

Note: this is a replace for PR (https://github.com/apache/flink/pull/7643) because it's corrupted.

bowenli86 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/ReadableWritableCatalog.java:
```diff
@@ -0,0 +1,175 @@
+	 */
+	void dropTable(ObjectPath objectName, boolean ignoreIfNotExists) throws TableNotExistException;
+
+
```
checkstyle: remove the empty line

bowenli86 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/exceptions/TableAlreadyExistException.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public class TableAlreadyExistException extends RuntimeException {
+
+	private static final String MSG = "Table (or view) $catalog.$table already exists.";
```
The MSG seems to be in scala style. Should be 'Table (or view) %s.%s already exists.'

bowenli86 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/exceptions/DatabaseNotExistException.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ */
+public class DatabaseNotExistException extends RuntimeException {
+	private static final String MSG = "Database $catalog.$database does not exist.";
```
The MSG seems to be in scala style. Should be 'Database %s.%s does not exist.'

bowenli86 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/exceptions/DatabaseAlreadyExistException.java:
```diff
@@ -0,0 +1,34 @@
+ * Exception for trying to create a database that already exists.
+ */
+public class DatabaseAlreadyExistException extends RuntimeException {
+	private static final String MSG = "Database $catalog.$database already exists.";
```
The MSG seems to be in scala style. Should be 'Database %s.%s already exists.'

bowenli86 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/ReadableWritableCatalog.java:
```diff
@@ -0,0 +1,175 @@
+		DatabaseNotExistException;
+
+	/**
+	 * Modifies an existing newView.
```
should it be "Modifies an existing view with newView"?

bowenli86 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/ReadableWritableCatalog.java:
```diff
@@ -0,0 +1,175 @@
+	/**
+	 * Modifies an existing newView.
+	 *
+	 * @param viewPath         Path of the newView to modify.
```
The javadoc seems to mix 'newView' with 'view', like "The new newView which replaces the existing newView.". May need some clean up

bowenli86 commented:
It LGTM now. Thanks!

sunjincheng121 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/ReadableWritableCatalog.java:
```diff
@@ -0,0 +1,167 @@
+	 *                          if set to true, do nothing.
+	 * @throws TableNotExistException if the given view does not exist
+	 */
+	void alterView(ObjectPath viewPath, CatalogView newView, boolean ignoreIfNotExists) throws TableNotExistException;
```
Do we need `renameView`?

xuefuz replied:
Makes sense. Added.

xuefuz commented:
> Thanks for the PR, I think catalog is very important for Table/SQL. Great job @xuefuz!
> 
> I quickly browsed the PR and left some comments.
> There are two other suggestions as follows:
> 
> * The overall PR does not have a test, it is recommended that  it better to add the  corresponding test case.
> * This PR just adds some interfaces, and there is no specific implementation and corresponding feature. This will confuse reviewer who have no context background, IMHO, each PR preferably has a specific function, of the function is large, it can be divided into multiple commits in one PR.
> 
> I am not sure whether my suggestion is reasonable or not, maybe @aljoscha and @twalthr can give us more advice.
> 
> What do you think?
> 
> Best,
> Jincheng

Thanks for the review, Jincheng! You have some good points.
First all, for those missing the contact, please check [FLIP-30.](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=100827366)

Secondly, this one does have a test and the reason is that this is mostly new interfaces, for which, tests are not really applicable.

Lastly, I don't believe that every merge needs to complete a feature or functionality. Incremental commits are preferable in my opinion because it makes review lighter and easier.

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/CatalogDatabase.java:
```diff
@@ -0,0 +1,45 @@
+ * Represents a database object in a catalog.
+ */
+public class CatalogDatabase {
+	// Property map for the database.
```
Interface is fine to me.

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/CatalogDatabase.java:
```diff
@@ -0,0 +1,45 @@
+	// Property map for the database.
+	private final Map<String, String> properties;
+
+	public CatalogDatabase() {
```
While for all meta objects there is a name, the name only matters in its enclosing parent. Thus, our design chose to not include name in the object definition. (Pls see FLIP-30.)

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/CatalogView.java:
```diff
@@ -0,0 +1,47 @@
+	 * Expanded query text takes care of this, as an example.
+	 *
+	 * <p>For example, for a view that is defined in the context of "default" database with a query "select * from test1",
+	 * the expanded query text might become "select `test1`.`name`, `test1`.`value` from `default`.`test1`", where
```
Good idea!

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/ReadableCatalog.java:
```diff
@@ -0,0 +1,126 @@
+	 * @return A list of the names of all tables and views in this database
+	 * @throws DatabaseNotExistException if the database does not exist
+	 */
+	List<String> listTables(String dbName) throws DatabaseNotExistException;
```
Different catalog may have specific ways to get tables or views only. However, it's commonly a practice that listTables() gives both while listViews() gives only the views.

sunjincheng121 replied:
I think we can add a TableType, such as `BASE_TABLE`, `VIEW`, and SQL server did in this way, please see the detail: https://chartio.com/resources/tutorials/sql-server-list-tables-how-to-show-all-tables/
welcome any feedback!

xuefuz replied:
Maybe you didn't have all the context, but this has been discussed previously and what we have now is what we concluded.

sunjincheng121 replied:
Yes, but do you mind telling me the real reason why we do not need let user list all of the tables(not views) directly?

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/ReadableCatalog.java:
```diff
@@ -0,0 +1,126 @@
+	 * @throws TableNotExistException if the table does not exist
+	 */
+	TableStats getTableStatistics(ObjectPath tablePath) throws TableNotExistException;
+
```
Good catch. It makes sense to remove it from here.

sunjincheng121 commented:
@flinkbot attention @aljoscha @twalthr

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/catalog/CommonTable.java:
```diff
@@ -0,0 +1,41 @@
+ * key-value pairs defining the properties of the table.
+ */
+public interface CommonTable {
+	/**
```
Same as above, name isn't managed by the class itself, but by the enclosing parent object.

bowenli86 commented on flink-table/flink-table-common/src/test/java/org/apache/flink/table/catalog/GenericInMemoryCatalogTest.java:
```diff
@@ -0,0 +1,581 @@
+/**
+ * Test for GenericInMemoryCatalog.
+ */
+public class GenericInMemoryCatalogTest {
```
if we keep ReadableWritableCatalog#renameDatabase(), we need to add some unit tests for it.

xuefuz replied:
ok.

xuefuz replied:
Actually I removed renameDatabase() from catalog APIs.

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:
```diff
@@ -0,0 +1,278 @@
+	public void renameDatabase(String name, String newName, boolean ignoreIfNotExists) throws DatabaseNotExistException {
+		if (databaseExists(name)) {
+			databases.put(newName, databases.remove(name));
+			tables.forEach((objectPath, table) -> {
```
not sure if I understand, but databases.remove(name) would remove it from the databases?

xuefuz replied:
Not sure if I I understand, but wouldn't databases.remove(name) in the code solves the problem?

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:
```diff
@@ -0,0 +1,278 @@
+	}
+
+	@Override
+	public void renameDatabase(String name, String newName, boolean ignoreIfNotExists) throws DatabaseNotExistException {
```
renameDatabase() was suggested by other review comments. I think it's fine to put it in place for completeness and for those that cannot support this, Unsupported operation exception may be thrown. I know there isn't a right/wrong answer, but it doesn't bother me much at least.

xuefuz replied:
Yeah. I guess there isn't right/wrong answer. This was added based on other review comment. It seems okay to me to keep this in place for completeness. For catalogs that cannot support it, an unsupported op ex may be thrown.

sunjincheng121 commented on flink-table/flink-table-common/src/test/java/org/apache/flink/table/catalog/CatalogTestUtil.java:
```diff
@@ -0,0 +1,76 @@
+		return new GenericCatalogTable(schema, stats, tableProperties);
+	}
+
+	public static void compare(GenericCatalogTable t1, GenericCatalogTable t2) {
```
Maybe change `compare` to `assertEquals` is more accurate? due to `compare` in general will return an INT.  What do you think?

sunjincheng121 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ObjectPath.java:
```diff
@@ -0,0 +1,88 @@
+
+	public ObjectPath(String databaseName, String objectName) {
+		checkNotNull(databaseName, "databaseName cannot be null");
+		checkNotNull(objectName, "objectName cannot be null");
```
What are the format restrictions for dataBaseName and objectName, such as special characters? e.g.: '.', duo to we using fullName.split("\\."). right?  So, it is better to add some check logic.
I had left the comment in pre-review.

xuefuz replied:
I added basic checks for databaseName and objectName (not-null, not-empty). However, I think the comprehensive checks should be left to parser. If API users provide ill-formed names, it's likely they will get errors in other APIs where this ObjectPath instance is used.

sunjincheng121 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ObjectPath.java:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	public static ObjectPath fromString(String fullName) {
+		String[] paths = fullName.split("\\.");
```
Please check the null case for `fullname`? 
Which I had left the comment in pre-review.

xuefuz replied:
Added checks for null and empty.

sunjincheng121 commented on flink-table/flink-table-common/src/test/java/org/apache/flink/table/catalog/GenericInMemoryCatalogTest.java:
```diff
@@ -0,0 +1,581 @@
+
+	@Before
+	public void setUp() {
+		catalog = new GenericInMemoryCatalog(db1);
```
You want to set `catalogName` right? So, we should change `db1` to `testCatalogName`.

sunjincheng121 commented on flink-table/flink-table-common/src/test/java/org/apache/flink/table/catalog/GenericInMemoryCatalogTest.java:
```diff
@@ -0,0 +1,581 @@
+		}
+	}
+
+	@AfterClass
```
Add @BeforeClass for  `open` logic?

xuefuz replied:
Sounds good.

sunjincheng121 commented on flink-table/flink-table-common/src/test/java/org/apache/flink/table/catalog/GenericInMemoryCatalogTest.java:
```diff
@@ -0,0 +1,581 @@
+	public void testCreateTable_DatabaseNotExistException() {
+		assertFalse(catalog.databaseExists(db1));
+
+		exception.expect(DatabaseNotExistException.class);
```
Add exception.expectMessage() to verify error message. Same for the tests below.

xuefuz replied:
Already added.

sunjincheng121 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:
```diff
@@ -0,0 +1,263 @@
+	}
+
+	@Override
+	public void renameView(ObjectPath viewPath, String newViewName, boolean ignoreIfNotExists)
```
Add test case for `renameView`?

xuefuz replied:
Added

sunjincheng121 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:
```diff
@@ -0,0 +1,278 @@
+/**
+ * A generic catalog implementation that holds all meta objects in memory.
+ */
+public class GenericInMemoryCatalog implements ReadableWritableCatalog {
```
Do we need `Generic` prefix? How about using `InMemoryCatalog` directly?  Or you already see the requirement for many kinds of `InMemoryCatalog`?

xuefuz replied:
The "generic" word used here to signify the catalog is for Flink metadata, which is similar to GenericHiveMetastoreCatalog. The naming comes from previous discussion with Timo.

sunjincheng121 replied:
Make sense to me.

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericCatalogDatabase.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * A generic catalog database implementation.
+ */
+public class GenericCatalogDatabase implements CatalogDatabase {
```
We overwrite them if the default one doesn't satisfy our needs. For a database, the most important part is its name, which isn't part of the object definition. I think it's fine if two DBs named differently happen to have the same definition (properties/comment). They are two different databases. Therefore, I believe reference equivalence is sufficient in this case.

sunjincheng121 replied:
I think it's better to overwrite. But I am fine if you do not want to do it.

xuefuz replied:
So far we haven't found a use case where we need to check non-reference equals. In fact we don't even find a case for comparison. Nevertheless, we can overwrite them if such use cases arrive in the future.

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericCatalogDatabase.java:
```diff
@@ -0,0 +1,59 @@
+	private String comment;
+
+	public GenericCatalogDatabase() {
+		this(new HashMap<>(), null);
```
Good point!

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericCatalogTable.java:
```diff
@@ -0,0 +1,73 @@
+	// Properties of the table
+	private final Map<String, String> properties;
+	// Statistics of the table
+	private TableStats tableStats = null;
```
"final" could be overkill, so I remove it.

sunjincheng121 replied:
I meant that we should add  `final`.  due to we only init it in the construct method. right?

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericCatalogTable.java:
```diff
@@ -0,0 +1,73 @@
+
+	@Override
+	public GenericCatalogTable copy() {
+		return new GenericCatalogTable(this.tableSchema, this.tableStats, this.properties);
```
Good catch!

sunjincheng121 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ReadableWritableCatalog.java:
```diff
@@ -0,0 +1,154 @@
+	 * @throws TableAlreadyExistException if table already exists and ignoreIfExists is false
+	 * @throws DatabaseNotExistException if the database in tablePath doesn't exist
+	 */
+	void createTable(ObjectPath tablePath, CatalogTable table, boolean ignoreIfExists)
```
It seems table in these APIs can be either table&view or table. For example, the `Table` in `getTable, renameTable, dropTable` means table and view, while the `Table` in `createTable, alterTable` means only table. It seems not very clear.

I think we have two options to solve the problem.
1.treat table as table and view for all APIs
2.treat table as the only table for all APIs. 
However, the first option may be more in line with the behavior of most standard databases.

xuefuz replied:
I think the two set of APIs are there for a reason.
1. getTable, renameTable, and dropTable only takes a name and nothing specific to either table or view.
2. on the contrary, createTable/View and alterTable/View takes a CatalogTable instance or  a CataView instance, and the implementations are expected to be different.

While it might seem confusing, it makes good sense if we look at the method signatures. Let me know if you still think this is confusing.

sunjincheng121 replied:
This involves an important concept, whether the Table contains (table and view), and if so, then I still recommend to keep a set of createTable/dropTable/renameTable/alterTable APIs, remove createView/alterView. any thoughts.?

xuefuz replied:
Well, createView/alterView takes different input than createTable/alterTable. If we mingle them, then the implementation of the API will be nasty because it will have to do something like if commonTable instance of CatalogTable then do xxx else if commonTable instanceof CatalogView, then do yyy. I'm not sure if we need a parity between view and table. There are things to be shared. let me go back to check how it's done in spark and presto to get some ideas there.

hequn8128 replied:
Sorry to jump into the discussion. Maybe it would be better if we can tell the differences directly and clearly from the method name. 

Currently, the method names are not trusted and we have to look at the signature to tell the differences. However, when we looking at the `List<String> listTables(String dbName)` method, it would again bring confusions. From the signature, a user will think both tables and views will be listed while actually only tables but not views are listed. 

I agree that the implementations of createTable and creatView are expected to be different, but I think it's ok to use one api for both of them. Take inmemory catalog as an example, we even don't need the `private helper method`. And we can use the `a private helper method` to solve the problem if the implementations are different. 

What do you guys think?

Best, 
Hequn

xuefuz replied:
Thanks all for the feedback. I don't think either method name or signature of the method or both can tell the whole story. That's the reason why we have API document, which is indispensable for API users. If one is still confused even after looking at the API documentation, then we have a problem. However, I don't think it's the case here. Plus, naming is very subjective and it can never be perfect!

Nevertheless, I take the point of merging APIs for regular tables and views into a single set, which is reflected in the latest update.

Let me know if there is anything to be further changed. Thanks.

sunjincheng121 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:
```diff
@@ -0,0 +1,319 @@
+			}
+		} else {
+			tables.put(viewPath, view.copy());
+		}
```
The implementations of createView() and createTable() are same. It's better to deduplicate the code. Same for alterView() and alterTable(). Again, I think maybe we can provide an api for both table and view, i.e., createTable(ObjectPath tablePath, CommonTable table, boolean ignoreIfExists) and alterTable(ObjectPath tablePath, CommonTable newTable, boolean ignoreIfNotExists). Table here means table and view.

xuefuz replied:
I can provide a private helper method that does the logic and both createView() and createTable() can call the helper method. Makes sense? The implementation here are the same because of inmemory nature. In general, they would be different, and CatalogTable and CatalogView are two different things. Thoughts?

sunjincheng121 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:
```diff
@@ -0,0 +1,319 @@
+	}
+
+	@Override
+	public List<String> listAllTables(String databaseName) throws DatabaseNotExistException {
```
`listAllTables`->`listTables`

xuefuz replied:
listTables() already exists in the class. see below.

sunjincheng121 replied:
I think listTables should contain both tables and views.

sunjincheng121 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:
```diff
@@ -0,0 +1,319 @@
+	}
+
+	@Override
+	public List<String> listTables(String databaseName) throws DatabaseNotExistException {
```
`listTables` -> `listTablesWithoutViews` or other better name.

xuefuz replied:
I used listAllTables to mean both, while listTables and listViews mean for each separately. This seems clear, especially with the javadoc. Would you rather use the long names?

sunjincheng121 replied:
Hmm, by design I am sure about `listTables` should contain both tables and views. If you do not like the name for listTableWithoutViews, we can just remove this method, and add it later.(when we got the nice name).

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ReadableCatalog.java:
```diff
@@ -0,0 +1,123 @@
+	 */
+	void setCurrentDatabase(String databaseName) throws DatabaseNotExistException;
+
+	// ------ databases ------
```
okay.

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:
```diff
@@ -0,0 +1,319 @@
+
+	}
+
+	// ------ databases ------
```
OKay.

dawidwys commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogDatabase.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Represents a database object in a catalog.
+ */
+public interface CatalogDatabase {
```
What is a database?

bowenli86 commented:
Hi all,

I believe @sunjincheng121 and @hequn8128 brought up valuable suggestions to avoid API name confusions. Xuefu also made very good points in consideration of API design and impl, and that javadoc should be the true for understanding APIs.

Previously I may be more affected by Hive's design given I've been working heavily on integrating Flink-Hive. @sunjincheng121 's concerns, if I understand correctly, may come from that these APIs will be used by not only SQL users but also Table API users, who may not have Hive backgrounds and thus easier to get confused. Thus I tried to step out of Hive context, and inspect these APIs from the perspective of their usage, as well as referencing MySQL, Postgres, Oracle, SQL Server, and Hive. Here are my thoughts:

On the reading side, view is always treated as a logical table. In queries (SELECT in standard SQL DML), view is table - 'FROM' clause is always "FROM x" rather than "FROM `TABLE/VIEW` x". It's planner's responsibility to process views specially. Meta commands as well, if with no extra params - "DESCRIBE" doesn't distinguish them; Listing tables usually goes in two syntax, "SHOW TABLES" and "SELECT * FROM meta", they return both tables and views, listing only views would be different commands or with extra params like "SHOW VIEWS" and "SELECT * FROM meta WHERE type='view'"

On the writing side, view is treated differently from table, given representations of view and table are a bit different (though they share some common fields). DDL, especially CREATE and ALTER,  are always requires specifying either `TABLE` or `VIEW` as "CREATE/ALTER `TABLE/VIEW` x". "DROP/RENAME" don't touch fields inside table and view, thus their impl behind the scene are usually the same, and therefore some databases choose to not require the `TABLE/VIEW` keyword, but I think it really depends on the developers. Since our devs feel strongly that it causes confusions, we can requires the keywords in our APIs and Flink SQL.

I think we should avoid design in which a SQL statement is translated into multiple catalog API calls or requires unnecessary extra processing. With that in mind, and also given the above conclusions (please correct me if there's anything above is wrong), I propose the following solution:`ReadableCatalog` APIs should treat views as tables by default if no extra params specified, thus `getTable()` and `listTables()` operate on both table and view, and we will have individual APIs as `listPhysicalTables()`, `listViews()`, and potentially `listMaterializedViews()` in the future. `ReadableWritableCatalog` APIs should treat views and tables differently, thus have create/alter/drop/rename APIs separately for view and table. E.g. dropTable() and dropView(), even though the two will very likely share the same code. We will also add clear javadoc and Flink documentations for all catalog APIs in a separate PR. This way, we can eliminate confusions and still maintain a 1 on 1 mapping between SQL statements and catalog APIs.

What do you think?

xuefuz commented:
I just pushed another round update, including:
1. Add "final" descriptor on tableStats variable in GenericCatalogTable.
2. removed existing listTables() and rename listAllTables() to listTables.
3. Impacted test.
Please take a look again and let me know if there is anything missing. Thanks.

dawidwys commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/exceptions/DatabaseNotExistException.java:
```diff
@@ -0,0 +1,35 @@
+ * Exception for trying to operate on a database that doesn't exist.
+ *
+ */
+public class DatabaseNotExistException extends RuntimeException {
```
nit: `DatabaseNotFoundException`? `NoSuchDatabaseException`? My personal feeling, but `DatabaseNotExistException` sounds strange to me.

xuefuz replied:
Yeah. I'm not particular on the naming but those names actually came from Flink exceptions.scala. I only migrated them to its java counterparts. Let me know if you think it's time to correct the names.

dawidwys replied:
Got it. I would be in favor of renaming them then. We are completely reworking the API, so I think we can fix the naming as well.

xuefuz replied:
I see. If you don't mind, I'd open a new JIRA on renaming the classes since this PR is almost ready and it didn't introduce the unfavorable names. Thanks.

twalthr replied:
I would also prefer fixing the names once and forall. Especially because we add `@PublicEvolving` annotations now.

xuefuz replied:
Sounds good. I create a new JIRA for fixing the names.

xuefuz replied:
Created https://issues.apache.org/jira/browse/FLINK-12255 to track the renaming.

dawidwys commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/exceptions/TableNotExistException.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Exception for trying to operate on a table (or view) that doesn't exist.
+ */
+public class TableNotExistException extends RuntimeException {
```
nit: `NoSuchTableException`? `TableNotFoundException`?

xuefuz replied:
Same as above.

dawidwys commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:
```diff
@@ -0,0 +1,271 @@
+	}
+
+	@Override
+	public void renameTable(ObjectPath tablePath, String newTableName, boolean ignoreIfNotExists)
```
Can't we relax this requirement and allow moving between databases? I know some catalogs don't support that, but others might.

xuefuz replied:
Yes, we can. However, I'd think pure renaming is probably more common than moving, which is the reason why it's here. For moving table between databases, we may provide a different API purely for that purpose rather than overloading it here, if that requirement ever comes. Any further thoughts?

dawidwys replied:
Ok, make sense.

wuchong commented:
Sorry to jump into the discussion. I find this PR has been discussed for a few days and we have some divergence in the `listTables()` method name. 

Actually, regarding to the [review process](https://flink.apache.org/reviewing-prs.html) (check item #2):

> In case of diverging opinions, the discussion should be moved to the respective Jira issue or to the dev mailing list and continued until consensus is reached. 

I would suggest to move the discussion to the mailing list, and invite people who were involved in the design doc (such as Timo, Shuyi) to make sure we have a consensus in the community. I'm sorry if this would delay the merging.

(Although we shouldn't continue the discussion in the PR, I still leave some thoughts here) 
Regarding to the naming, IMO the interface of catalog do not have to match SQL DDL. SQL standard doesn't define the catalog interface. So I prefer to separate methods for listing tables and views which is more clear to devs.

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogView.java:
```diff
@@ -0,0 +1,47 @@
+	 *
+	 * @return the view definition in expanded text.
+	 */
+	String getExpandedQuery();
```
Functionally, getExpandedQuery is enough. getOriginalQuery() is there for logging and debugging purpose, which I think useful.

dawidwys replied:
Will both of them be stored in catalog? Does `originalQuery` makes sense when read from catalog, if settings of the session that reads that entry may have different default catalog than the one that wrote that entry.

I don't know I feel like this might introduce unnecessary problems. I would prefer to operate only with the "expanded" query version.

xuefuz replied:
Yes. Both of them are stored in the catalog at least for Hive. However, when the view is used for query, only expanded query is used. So, the problem you referred to can be avoided.

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogDatabase.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Represents a database object in a catalog.
+ */
+public interface CatalogDatabase {
```
As can be seen throughout the design and code here, every meta class has properties. As to database, it can have properties like created_at, created_by, last_ddl_time, etc. Maybe not every catalog need it, but I think in general we need to support those on the API level.

dawidwys replied:
Do we know of any catalogs that would need that concept? I don't know I don't see any benefits in having those classes right now. Those properties you mentioned are system properties, which should be auto generated, are we going to do that for all catalogs?

I really feel that we do not have any use cases for those classes. Could we do the same as with "moving tables between databases" and introduce it once we know what do we need them for? It's always easier to introduce new API than remove old one.

sunjincheng121 replied:
In the Catalog, we can have multiple `CatalogDatabases`. There are multiple `CatalogTables` and `CatalogViews` in each `CatalogDatabase`. I think the meaning of the `CatalogDatabase` is the management of the CommonTable. The management relationship here is a bit special. For example, the `CatalogDatabase` and the CommonTable are the members of XXXReadableCatalog, and members are flat managed. They are dependent on `ObjectPath`. For example, `ObjectPath` contains the databaseName attribute. This management method is similar to the way that spark integrates hive, and it is flat management. Please see: https://github.com/apache/spark/blob/bbbe54aa79c0f6b66e3f3ac34515cc096beb5730/sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala#L237

Feel free to correct me if there is anything I misunderstood. :-)

dawidwys replied:
I totally get the multi stage referencing and paths and I am totally for them, but we don't need the class `CatalogDatabes` for that. I just don't see a use cases for having an independent `CatalogDatabase` representation. 

When and what for would we need to do `Catalog.getDatabase()`?

sunjincheng121 replied:
From the points of view, the user may want to see all databases in the catalog by `listDatabases`. and if user wants to see the detail for database info, such as properties, will call `getDatabase(..)`.

xuefuz replied:
> Hi @sunjincheng121 , @xuefuz , @bowenli86
> 
> Regarding to the `listTables()`, I'm fine with current approach that `listTables` returns tables and views if you all prefer it. I don't have a strong opinion on this.
> 
> Regarding moving to the mailing list, my initial intention is involving more people into the discussion when we have diverging opinions on API, but it seems that we have a consensus on this now. I agree with @sunjincheng121 that if we think the API change is minor, I think it's fine to keep the discussion in PR. But I would suggest to update the FLIP according to the latest change after the pull request is merged.
> 
> So, I'm +1 to merge this after Dawid's comments addressed.

Yes, I will update the FLIP to reflect what we have agreed upon. Thanks.

xuefuz commented:
> Sorry to jump into the discussion. I find this PR has been discussed for a few days and we have some divergence in the `listTables()` method name.
> 
> Actually, regarding to the [review process](https://flink.apache.org/reviewing-prs.html) (check item #2):
> 
> > In case of diverging opinions, the discussion should be moved to the respective Jira issue or to the dev mailing list and continued until consensus is reached.
> 
> I would suggest to move the discussion to the mailing list, and invite people who were involved in the design doc (such as @twalthr , @suez1224 ) to make sure we have a consensus in the community. I'm sorry if this would delay the merging.
> 
> (Although we shouldn't continue the discussion in the PR, I still leave some thoughts here)
> Regarding to the naming, IMO the interface of catalog do not have to match SQL DDL. SQL standard doesn't define the catalog interface. So I prefer to separate methods for listing tables and views which is more clear to devs.

@wuchong Thanks for sharing your thoughts. Agreed with you on the review process. Unfortunately there was a discontinuity between the design and review, which triggered some discussions here. On the other hand, w/o active review, those design discussions may never come.

Nevertheless, I don't think we need to resolve every disagreement before we proceed. Per the listTable issue you mentioned, if we agree on the latest listTables() and listViews(), then we should include them in this PR while leaving out the controversial one (listTableWithoutViews()), for which there can be different solutions and it's not a show-stopper. We can come back later to add that when a consensus is reached.

Let me know if this makes sense. Thanks.

sunjincheng121 commented:
Thank for the comments and reply @wuchong @dawidwys  @xuefuz @bowenli86 :-)

I agree that catalog API is different from SQL DDL. But I think it shares big common with SQL APIs and It would be good if we can keep the catalog API align with SQL, especially for some very common methods like `listTables`.

In the two approaches I proposed to unify the API, I still like listTables to return tables and views together. (I agree with the current(7efb860) PR changes about `listTables`).

Regarding moving to the mailing list, If most of the changes in the PR have been agreed, and there are only a few different opinions, such as: `listTables` not have a consensus, it is an efficient way to discuss in the PR and reach a consensus. As for the review, I don't think we need to move the current discussion into the mailing list.

- We have already gone over the PR very carefully. And it comes very close to the merging.
- I don't think we should bring everything to the dev mailing list but a discussion on the Jira ticket would be also good(just like the quote from jark's reply).

From the points of my view, the PR looks good to me, after @xuefuz addresses the comments form @dawidwys.

What do you think?

Best,
Jincheng

bowenli86 commented:
Re Jark's comment, I agree with Jincheng. My point is that catalog APIs should be a super set of SQL, each plain SQL syntax is better to be mapped to one catalog API in design, and beside those, catalog APIs can have more methods.

With Dawid's comments being taken care of, I think this PR will be in a good shape. Thanks @sunjincheng121 for pushing this forward

wuchong commented:
Hi @sunjincheng121 , @xuefuz , @bowenli86 

Regarding to the `listTables()`, I'm fine with current approach that `listTables` returns tables and views  if you all prefer it. I don't have a strong opinion on this.

Regarding moving to the mailing list, my initial intention is involving more people into the discussion when we have diverging opinions on API, but it seems that we have a consensus on this now. I agree with @sunjincheng121 that if we think the API change is minor, I think it's fine to keep the discussion in PR. But I would suggest to update the FLIP according to the latest change after the pull request is merged.

So, I'm +1 to merge this after Dawid's comments addressed.

xuefuz commented:
> Hi @sunjincheng121 , @xuefuz , @bowenli86
> 
> Regarding to the `listTables()`, I'm fine with current approach that `listTables` returns tables and views if you all prefer it. I don't have a strong opinion on this.
> 
> Regarding moving to the mailing list, my initial intention is involving more people into the discussion when we have diverging opinions on API, but it seems that we have a consensus on this now. I agree with @sunjincheng121 that if we think the API change is minor, I think it's fine to keep the discussion in PR. But I would suggest to update the FLIP according to the latest change after the pull request is merged.
> 
> So, I'm +1 to merge this after Dawid's comments addressed.

Yes, I will update the FLIP to reflect what we agreed upon here.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogDatabase.java:
```diff
@@ -0,0 +1,38 @@
+	/**
+	 * Get a map of properties associated with the database.
+	 */
+	Map<String, String> getProperties();
```
Add more very common methods, I would suggest to at at least `getDescription(): Optional<String>` (usually one sentence for the CLI listing) and `getDetailedDescription(): Optional<String>` (a lot of text).

xuefuz replied:
Hi Timo, Thanks for the feedback.

By "description", do you mean "comment" that is given for a database or table? It's fine to add getComment() here. However, typical databases only allow user to give just one common. Therefore, I'm not sure if we need getDetailedDescription() and how one might implement it. Any further thoughts?

twalthr replied:
I chose the naming `getDescription()` and `getDetailedDescription()` to be in sync with UDFs that will also receive this interfaces. I would also be fine with just `getDescription()`. However, I think it would be helpful to have a read-only method that returns a catalog dependent explanation string. For Hive, such a string could contain LOCATION, USER, ROLES etc. Information that might be interesting for users during a CLI session that is not part of the writable comment field.

xuefuz replied:
Okay, sounds good.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CommonTable.java:
```diff
@@ -0,0 +1,47 @@
+ * CommonTable is the common parent of table and view. It has a map of
+ * key-value pairs defining the properties of the table.
+ */
+public interface CommonTable {
```
Prefix this with `CatalogCommonTable`.

twalthr replied:
or should we call it `CatalogBaseTable`?

xuefuz replied:
CatalogBaseTable seems good. Thanks.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogView.java:
```diff
@@ -0,0 +1,47 @@
+public interface CatalogView extends CommonTable {
+
+	/**
+	 * Original text of the view definition.
```
Add a comment that this method also preserves formatting.

xuefuz replied:
OK.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CommonTable.java:
```diff
@@ -0,0 +1,47 @@
+	 * Get a deep copy of the CommonTable instance.
+	 * @return a copy of the CommonTable instance
+	 */
+	CommonTable copy();
```
Same comment as for database. Please add methods for a description and detailed description.

xuefuz replied:
Will do when comment on database is clarified.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericCatalogDatabase.java:
```diff
@@ -0,0 +1,60 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog;
```
This belongs to `api-java`. `common` should only contain extension points.

xuefuz replied:
Sounds good.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ReadableCatalog.java:
```diff
@@ -0,0 +1,116 @@
+ * This interface is responsible for reading database/table/views/UDFs from a registered catalog.
+ * It connects a registered catalog and Flink's Table API.
+ */
+public interface ReadableCatalog {
```
All methods in this class can throw a generic `CatalogException` if the communication fails or the API is not used as intended.

xuefuz replied:
Hi Timo, CatalogException you mentioned sounds like a RuntimeException that don't need to be declared. However, I noticed that the currently declared exceptions inherited also RuntimeException, which should be just exception. I'd suggest we change those as checked exception and reserve the generic CatalogException as a RuntimeException that might be thrown in all catalog APIs. How does this sound to you?

twalthr replied:
Yes, checked exceptions make sense for all catalog methods. I'm fine with keeping `CatalogException` a runtime exception but I would add it to all catalog APIs to remind implementers that communication to catalogs is always error-prone.

xuefuz replied:
Sounds good. I will have CatalogException declared in those APIs.

xuefuz commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ReadableCatalog.java:
```diff
@@ -0,0 +1,116 @@
+	 *
+	 * @param databaseName		Name of the database
+	 */
+	boolean databaseExists(String databaseName);
```
I noticed databaseExists() or schemaExists() are fairly commonly used in other database systems. On the other hand, word "contains" sounds more like a container, which isn't exactly what a catalog is about. Let me know if you're strong on this. Thanks.

twalthr replied:
From a Java world perspective, methods start with a verb. So people expect `has/contains/is`. But I don't have a strong opinion here.

xuefuz commented:
Hi Timo, Thanks for your feedback. I have responded to each your comments and some of them need your further thoughts or clarifications. Let me know if I miss anything. Thanks.

xuefuz commented:
PR is updated based on above discussions:

1. Changed existing exceptions to checked and added a runtime exception, CatalogException
2. Declared CatalogException in the APIs
3. Added getDescription() and getDetailedDescription() in table/database API
4. Move generic catalog and related class to flink-table-api-java.
5. Renamed CommonTable to CatalogBaseTable.

Please take another look and let me know if additional changes are needed. Thanks.

sunjincheng121 commented:
@xuefuz thanks for the update!

+1 to merged.

@flinkbot approve all

Best,
Jincheng

sunjincheng121 commented:
Mergint...
